xhype/xhype/src/decode.rs

```diff
@@ -0,0 +1,265 @@
+#[allow(unused_imports)]
+use log::*;
+
+type MemAccessFn = fn(&VCPU, &mut GuestThread, usize, &mut u64, u8, bool) -> Result<(), Error>;
+
+#[derive(Debug)]
+pub struct X86Decode {
+    prefix_sz: u8,
+    opcode_sz: u8,
+    modrm_sib_sz: u8,
+    imm_sz: u8,
+    operand_bytes: u8,
+    address_bytes: u8,
+    has_modrm: bool,
+    is_store: bool,
+    rex_r: bool,
+    rex_x: bool,
+    rex_b: bool,
+}
+
+impl Default for X86Decode {
+    fn default() -> Self {
+        X86Decode {
+            prefix_sz: 0,
+            opcode_sz: 1,
+            modrm_sib_sz: 0,
+            imm_sz: 0,
+            operand_bytes: 4,
+            address_bytes: 8,
+            has_modrm: true,
+            is_store: false,
+            rex_r: false,
+            rex_x: false,
+            rex_b: false,
+        }
+    }
+}
+
+fn decode_prefix(insn: &[u8], decode: &mut X86Decode) {
+    let mut prefix_sz = 0;
+    for byte in insn.iter() {
+        if *byte == 0x66 {
+            /* Operand-size override prefix */
+            /* Ignore 0x66 if REX.W changed us to 8 bytes (64).
+             * Though we should only see 0x66 before REX.W.
+             *
+             * If this was handling 32 bit code but with cs.d clear
+             * (default 16), 66 should set us to 4 bytes. */
+            if decode.operand_bytes == 4 {
+                decode.operand_bytes = 2;
+            }
+        } else if *byte == 0x67 {
+            /* Address-size override prefix */
+            decode.address_bytes = 4;
+        } else if *byte & 0xf0 == 0x40 {
+            /* REX.* */
+            if *byte & 0x08 != 0 {
+                decode.operand_bytes = 8;
+            }
+            if *byte & 0x04 != 0 {
+                decode.rex_r = true;
+            }
+            if *byte & 0x02 != 0 {
+                decode.rex_x = true;
+            }
+            if *byte & 0x01 != 0 {
+                decode.rex_b = true;
+            }
+        } else {
+            break;
+        }
+        prefix_sz += 1;
+    }
+    decode.prefix_sz = prefix_sz;
+}
+
+fn get_modrm(insn: &[u8], decode: &X86Decode) -> Result<u8, Error> {
+    if decode.has_modrm {
+        Ok(insn[(decode.prefix_sz + decode.opcode_sz) as usize])
+    } else {
+        Err("No modrm".to_string())?
+    }
+}
+
+fn modrm_get_reg(insn: &[u8], decode: &X86Decode) -> Result<u8, Error> {
+    let modrm = get_modrm(insn, decode)?;
+    let reg = (modrm >> 3) & 7;
+    match decode.address_bytes {
+        2 => {
+            error!("decode: had 2 address bytes");
+            Ok(reg)
+        }
+        4 | 8 => {
+            if decode.rex_r {
+                Ok(reg + 8)
+            } else {
+                Ok(reg)
+            }
+        }
+        _ => {
+            let msg = format!("decode: had {} address bytes", decode.address_bytes);
+            error!("{}", msg);
+            Err(msg)?
+        }
+    }
+}
+
+fn modrm_sib_bytes_16(mod_: u8, rm: u8) -> u8 {
+    let mut ret = 1;
+    match mod_ {
```

Comment:
I'm curious, could this match be written as 
match (mod_, rm)
0, 6 => ret += 2;
0, _ =>{}
1, 4 => ret += 2;
1, _ => ret += 1;
2, _=> ret += 2;
_, _ => {}

it almost seems more readable that way

Reply:
I updated